### PR TITLE
Logging

### PR DESCRIPTION
**Pull request by hanars**
This PR reformats our logs to optimize them for monitoring, and adds detailed logging for all http requests and database transactions. The big changes are:

- log everything as a detailed json instead of a string. This is done via updates to the `LOGGING` configuration in `settings.py` and the custom formatter defined in `seqr.utils.logging_utils.JsonLogFormatter`. Samples of what this look like are:
```
{"timestamp": "2020-10-26 20:56:05,493", "severity": "INFO", "message": "Fetched 3 Samples records from airtable"}

{"timestamp": "2020-10-26 20:56:07,587", "severity": "WARNING", "httpRequest": {"requestMethod": "POST", "requestUrl": "http://testserver/api/search_context", "status": 400, "responseSize": 72, "userAgent": null, "remoteIp": "127.0.0.1", "referer": null, "protocol": "HTTP/1.1"}, "requestBody": {"searchHash": "d380ed0fd28c3127d07a64ea2ba907d7"}, "user": "test_user_no_staff@test.com"}

{"timestamp": "2020-10-26 20:56:08,741", "severity": "INFO", "message": "Updated Family F000003_3", "dbUpdate": {"dbEntity": "Family", "entityId": "F000003_3", "updateType": "update", "updateFields": ["pedigree_image"]}}
```

- add detailed logging for every http request. Before, we were using django's default server logging but we switched to a custom logging middleware (`LogRequestMiddleware`) because the server logging doesn't have access to the post body, which we are required to include in our logs for FISMA compliance
- add detailed logging for all database operations. There is unfortunately no middleware-esque equivalent to capture all database events, so this is done by adding logging to helper methods, and then doing all database operations using those helpers. This means the diff is pretty big, although there is minimal behavior change

### CODE REVIEW

File: seqr/utils/middleware.py [hunk 1]
```diff
@@ -40,3 +41,38 @@ def process_exception(request, exception):
...
+            'referer': request.META.get('HTTP_REFERER'),
+            'protocol': request.META.get('SERVER_PROTOCOL'),
+        }
+        request_body = None
+        try:
+            if request.body:
+                request_body = json.loads(request.body)
+                # TODO update settings in stackdriver so this isn't neccessary
+                if 'password' in request_body:
+                    request_body['password'] = '***'
```
Comment (hanars) [inline seqr/utils/middleware.py:66]: Codacy found an issue: [Possible hardcoded password: '***'](https://app.codacy.com/manual/hanars/seqr/pullRequest?prid=6404089)

**Comment by hanars** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- seqr/views/utils/dataset_utils.py  1
- seqr/views/utils/json_to_orm_utils.py  1
- seqr/utils/middleware.py  2
- seqr/utils/logging_utils.py  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/hanars/seqr/pullRequest?prid=6404089&bid=20758730)

**Comment by hanars** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- seqr/views/utils/dataset_utils.py  1
- seqr/utils/logging_utils.py  6
- seqr/views/utils/json_to_orm_utils.py  1
- seqr/utils/middleware.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/hanars/seqr/pullRequest?prid=6404089&bid=20758730)

File: seqr/utils/middleware.py (hunk 1 above)
Comment (hanars) [inline seqr/utils/middleware.py:66]: Codacy found an issue: [Possible hardcoded password: '***'](https://app.codacy.com/manual/hanars/seqr/pullRequest?prid=6404089)

Comment (hanars) [reply seqr/utils/middleware.py:66]: this is okay to ignore

**Comment by hanars** [comment]
The codacy issue is okay to ignore

**Comment by hanars** [comment]
passing tests: https://travis-ci.org/github/broadinstitute/seqr/builds/739111049

**Comment by ShifaSZ** [comment]
I have an overall comment that the logs shouldn't contain any secrets. I have a concern about logging the POST body because the body could have some secrets such as keys or tokens.

File: seqr/utils/logging_utils.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        else:
+            log_json['message'] = record.message
+
+        if getattr(record, 'user', None) and record.user.is_authenticated():
+            log_json['user'] = record.user.email
+
+        if hasattr(record, 'db_update'):
+            log_json['dbUpdate'] = record.db_update
+
+        return json.dumps(log_json)
```
Comment (ShifaSZ) [inline seqr/utils/logging_utils.py:24]: Need an extra line.

**Comment by hanars** [comment]
> I have an overall comment that the logs shouldn't contain any secrets. I have a concern about logging the POST body because the body could have some secrets such as keys or tokens.

I agree, but this is a non-negotiable requirement for FISMA. I raised these concerns with David Bernick and was informed we still have to log the post body. Realistically, keys and tokens should always be sent as headers and never in the body, so the only real risk is passwords during login password reset

**Comment by ShifaSZ** [comment]
One question is: why don't you use the `django.db.backends` logger?

**Comment by hanars** [comment]
> One question is: why don't you use the `django.db.backends` logger?

1) because that logs the raw sql statement that is executed which we aren't really interested in, because its hard to parse from the text of a sql statement exactly what happened
2) because theres no way to include the associated user using that logger, and thats arguably the most important thing in an audit log

**Review by ShifaSZ — APPROVED** [review]
Awesome overall. There is no critical issue found other than a few questions.

File: seqr/management/commands/lift_project_to_hg38.py [hunk 3]
```diff
@@ -81,7 +82,7 @@ def handle(self, *args, **options):
                 ))
 
         # Lift-over saved variants
-        _update_variant_samples(matched_sample_id_to_sample_record, elasticsearch_index)
+        _update_variant_samples(matched_sample_id_to_sample_record, None, elasticsearch_index)
```
[the remark this answers is not in the document]
Comment (hanars) [reply seqr/management/commands/lift_project_to_hg38.py:85]: the user is only used for logging and since logs for ad-hoc manage commands don't actually go to stackdriver/ anywhere persistent theres no reason to add unnecessary code complexity

File: seqr/views/apis/locus_list_api.py [hunk 4]
```diff
@@ -114,6 +112,9 @@ def add_project_locus_lists(request, project_guid):
     for locus_list in locus_lists:
         locus_list.projects.add(project)
         locus_list.save()
+    logger.info('Updated Project {}'.format(project_guid), extra={'user': request.user, 'db_update': {
+        'dbEntity': 'Project', 'entityId': project_guid, 'updateType': 'update', 'updateFields': ['locus_lists'],
+    }})
```
[the remark this answers is not in the document]
Comment (hanars) [reply seqr/views/apis/locus_list_api.py:117]: Yea, its because the project is being added the the `LocusList` model, but we also want to log that that means the project itself changes, because there is a many-to-many relationship between them. I can at least use a helper method for generating the log message though so I will do that

File: seqr/views/apis/users_api.py [hunk 5]
```diff
@@ -78,9 +82,8 @@ def set_password(request, username):
         return create_json_response({}, status=400, reason='Password is required')
 
     user.set_password(request_json['password'])
-    user.first_name = request_json.get('firstName') or ''
-    user.last_name = request_json.get('lastName') or ''
-    user.save()
+    update_model_from_json(user, _get_user_json(request_json), user=user, updated_fields={'password'})
+    logger.info('Set password for user {}'.format(user.email), extra={'user': user})
```
[the remark this answers is not in the document]
Comment (hanars) [reply seqr/views/apis/users_api.py:86]: It technically doesn't but it seemed like a good idea to have extra logs related to login and logout and user management in general since those are the ones that are regularly called out in the FISMA compliance docs

File: seqr/views/apis/variant_search_api.py [hunk 6]
```diff
@@ -437,21 +441,18 @@ def create_saved_search_handler(request):
...
     except MultipleObjectsReturned:
         # Can't create a unique constraint on JSON field, so its possible that a duplicate gets made by accident
         dup_searches = VariantSearch.objects.filter(
             search=request_json,
             created_by=request.user,
         ).order_by('created_date')
         saved_search = dup_searches[0]
-        for search in dup_searches:
+        for search in dup_searches[1:]:
             search.delete()
```
[the remark this answers is not in the document]
Comment (hanars) [reply seqr/views/apis/variant_search_api.py:454]: no these models would have been created less than a second earlier, its just an issue with integrity/concurrent requests and they aren''t actually "real" models

[the remark this answers is not in the document]
Comment (hanars) [reply seqr/views/apis/variant_search_api.py:454]: actually I can just use the `bulk_delete` helper for this, so it will get logged

File: seqr/utils/middleware.py (hunk 1 above)
Comment (hanars) [inline seqr/utils/middleware.py:66]: Codacy found an issue: [Possible hardcoded password: '***'](https://app.codacy.com/manual/hanars/seqr/pullRequest?prid=6404089)

**Comment by hanars** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- seqr/utils/middleware.py  2
- seqr/utils/logging_utils.py  6
- seqr/views/utils/json_to_orm_utils.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/hanars/seqr/pullRequest?prid=6404089&bid=20758730)
